### DISCUSSION TITLE
[question] Manual execution

### DISCUSSION THREAD

**Discussion by sieskei** [xCuri0/ReBarUEFI#54] [General]
Hello, I have an interesting case. I have an Asus M16 2021 laptop and I'm using an eGPU (RX 6800 XT) through an Oculink x4 Gen 4.0 connection. It's connected via the NVMe port and works great, but without resizable BAR (re-bar) support.

I managed to enable resizable BAR support using setup_var.efi, but it only works for the built-in GPU (dGPU, 3060). I even disabled the dGPU and iGPU through the BIOS, but it still doesn't work.

An interesting fact is that in Linux, I can boot with pci=realloc and set bar2 to 16GB. I can even dynamically resize it at runtime using resource2_resize.

So, my thought is the following: Is it possible to create this an EFI tool that can be invoked manually before booting into Windows?

It would be great for people like me who don't have the option to flash the entire BIOS.

Thank you in advance!
Miro.

**Comment by xCuri0** [comment]
@sieskei this module works by hooking `PreprocessController` which gets called by the PciBus module during enumeration. unless you can figure out how to re-enumerate PCI I don't think so.

using `kexec` to boot windows from linux is possible and has been done before.

see https://github.com/osresearch/safeboot-loader. you will need to setup a minimal kernel + initrd which does the bar resizing and then boots into windows

**Reply by sieskei** [reply]
Hello, thank you for your response!

Yes, I understand that what you have done works in a different way.
I was just wondering if there is an easy way to implement the logic of the Linux kernel for PCI realloc.

I also had the idea of booting Linux first and then transitioning to Windows, but it seems a bit overkill.

BTW Why does resizable BAR only work for the dgpu (rxt 3060) (PEG10), even after disabling the entire PCIe Root Controller 2? The external graphics card is connected via PCIe Root Controller 1 (PEG60). Is there a chance to reallocate this controller through a DSDT/SSDT hack?

**Reply by xCuri0** [reply]
@sieskei it's probably the bios being written to only resize stuff on that port don't think it can be easily fixed.

Using Linux to boot windows is overkill yeah but I think it's the best option.

 But I did try another method before where I would resize and assign BAR for GPU before Windows boots with an EFI program, it worked but for some reason the driver showed resizable bar as disabled (GPU-Z showed it working 8GB) and I didn't see any performance benefit. I guess I did something wrong in it

You can find that in the first commit of this repo

**Reply by sieskei** [reply]
@xCuri0 thank you!

Does the fact that the BIOS does not support ReBAR on that port mean that even if DXE is added, it still won't work?

The kexec option seems interesting, and I would like to try it. However, I'm having trouble finding much available information suitable for my level of experience and knowledge in Linux.

**Reply by xCuri0** [reply]
@sieskei dxe should work if the BIOS doesn't have anything preventing it from being modified

**Reply by bannhead** [reply]
> @xCuri0 thank you!
> 
> Does the fact that the BIOS does not support ReBAR on that port mean that even if DXE is added, it still won't work?
> 
> The kexec option seems interesting, and I would like to try it. However, I'm having trouble finding much available information suitable for my level of experience and knowledge in Linux.

Hi @sieskei, were you able to put together a working bootloader to start the notebook with nvme activated rebar in Windows? I'm in the same situation as you, with a Dell laptop that can't change the bios

**Reply by xCuri0** [reply]
@bannhead have you tried setting up safeboot-loader ?

you just need to set it up and then use sysfs (resource#_resize) to resize bar before chainloading windows. you may need to edit DSDT too if your laptop doesn't have 4g decoding

**Reply by sieskei** [reply]
Hello, I haven't tried it yet because I spent a long time attempting to flash the BIOS, but it seems that it won't work without a programmer.

This bootloader looks interesting. Is it still supported? In the description, it mentions that it uses kernel version 5. Are the features related to ReBar available in versions 6 and above?

**Reply by xCuri0** [reply]
@sieskei afaik kernel version 5 has them but doesn't expose them through sysfs. But it shouldn't be too hard to write a simple kernel driver that resizes it
